File: merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> merge_sort(const vector<int> &data)
{
    if (data.size() <= 1) {
        return data;
    }

    int middle = data.size() / 2;
    vector<int> left(data.begin(), data.begin()+middle);
    vector<int> right(data.begin()+middle, data.end());

    left = merge_sort(left);
    right = merge_sort(right);

    vector<int> result(data.size());

    merge(left.begin(), left.end(),
          right.begin(), right.end(),
          result.begin());

    return result;
}

int main()
{
    int n, x;
    vector<int> a;

    scanf("%d", &n);

    for (int i = 0; i < n; ++i)
    {
        scanf("%d",&x);
        a.push_back(x);
    }

    vector<int> result = merge_sort(a);
    for (int i = 0; i < result.size(); ++i)
    {
        printf("%d\n", result[i]);
    }

    return 0;
}

File: merge_sort.py
def simpleMerge(a, b):
    total = len(a) + len(b)
    j, k = 0, 0

    c = list()

    for i in range(total):
        if (k == len(b) or (j < len(a) and a[j] < b[k])):
            c.append(a[j])
            j += 1
        else:
            c.append(b[k])
            k += 1

    return(c)

def mergeSort(data):
    if (len(data) <= 1):
        return(data)

    middle = int(len(data) / 2)
    left = mergeSort(data[0 : middle]) # O data de 0 até middle
    right = mergeSort(data[middle : len(data)]) # O data de middle até o final

    result = simpleMerge(left, right)
    return(result)

a = list()
n = int(input())

for i in range(n):
    a.append(int(input()))

result = mergeSort(a)

for i in result:
    print(i)
